native/dtrevc3.go

```diff
@@ -0,0 +1,862 @@
+	switch side {
+	default:
+		panic(badSide)
+	case lapack.RightEigVec, lapack.LeftEigVec, lapack.RightLeftEigVec:
+	}
+	switch howmny {
+	default:
+		panic(badHowMany)
+	case lapack.AllEigVec, lapack.AllEigVecMulQ, lapack.SelectedEigVec:
+	}
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case len(work) < lwork:
+		panic(shortWork)
+	case lwork < max(1, 3*n) && lwork != -1:
+		panic(badWork)
+	}
+	if lwork != -1 {
+		if howmny == lapack.SelectedEigVec {
+			if len(selected) != n {
+				panic("lapack: bad selected length")
+			}
+			// Set m to the number of columns required to store the
+			// selected eigenvectors, and standardize the slice
+			// selected.
+			for j := 0; j < n; {
+				if j == n-1 || t[(j+1)*ldt+j] == 0 {
+					if selected[j] {
+						m++
+					}
+					j++
+					continue
+				}
+				if selected[j] || selected[j+1] {
+					selected[j] = true
+					selected[j+1] = false
+					m += 2
+				}
+				j += 2
+			}
+		} else {
+			m = n
+		}
+		if m > mm {
+			panic("lapack: insufficient number of columns")
+		}
+		checkMatrix(n, n, t, ldt)
+		if (side == lapack.RightEigVec || side == lapack.RightLeftEigVec) && m > 0 {
+			checkMatrix(n, m, vr, ldvr)
+		}
+		if (side == lapack.LeftEigVec || side == lapack.RightLeftEigVec) && m > 0 {
+			checkMatrix(n, m, vl, ldvl)
+		}
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		work[0] = 1
+		return
+	}
+
+	nb := impl.Ilaenv(1, "DTREVC", string(side)+string(howmny), n, -1, -1, -1)
+
+	// Quick return in case of a workspace query.
+	if lwork == -1 {
+		work[0] = float64(n + 2*n*nb)
+		return m
+	}
+
+	const (
+		nbmin = 8
+		nbmax = 128
+	)
+
+	// Use blocked version of back-transformation if sufficient workspace.
+	// Zero-out the workspace to avoid potential NaN propagation.
+	if howmny == lapack.AllEigVecMulQ && lwork >= n+2*n*nbmin {
+		nb = min((lwork-n)/(2*n), nbmax)
+		impl.Dlaset(blas.All, n, 1+2*nb, 0, 0, work, 1+2*nb)
+	} else {
+		nb = 1
+	}
+
+	// Set the constants to control overflow.
+	ulp := dlamchP
+	smlnum := float64(n) / ulp * dlamchS
+	bignum := (1 - ulp) / smlnum
+
+	// Split work into a vector of column norms and an n×2*nb matrix b.
+	norms := work[:n]
+	ldb := 2 * nb
+	b := work[n : n+n*ldb]
+
+	// Compute 1-norm of each column of strictly upper triangular part of T
+	// to control overflow in triangular solver.
+	norms[0] = 0
+	for j := 1; j < n; j++ {
+		norms[j] = 0
+		for i := 0; i < j; i++ {
+			norms[j] += math.Abs(t[i*ldt+j])
+		}
+	}
+
+	bi := blas64.Implementation()
+
+	var (
+		x [4]float64
+
+		iv int // Index of column in current block.
+		is int
+
+		// ip is used below to specify the real or complex eigenvalue:
+		//  ip == 0, real eigenvalue,
+		//        1, first  of conjugate complex pair (wr,wi),
+		//       -1, second of conjugate complex pair (wr,wi).
+		ip        int
+		iscomplex [nbmax]int // Stores ip for each column in current block.
+	)
+
+	if side == lapack.LeftEigVec {
```

Comment:
Yeah, me too. I did it to reduce the indentation, it made the painful work a tiny bit more convenient. I can remove it now, I really have no preference.